font/font.go

```diff
@@ -0,0 +1,304 @@
+type FontExtents struct {
+	// Ascent is the distance that the text
+	// extends above the baseline.
+	Ascent Length
+
+	// Descent is the distance that the text
+	// extends below the baseline.  The descent
+	// is given as a positive value.
+	Descent Length
+
+	// Height is the distance from the lowest
+	// descending point to the highest ascending
+	// point.
+	Height Length
+}
+
+type FontFace struct {
+	Font Font
+	Face *opentype.Font
+}
+
+// Name returns a fully qualified name for the given font.
+func (f *FontFace) Name() string {
+	return f.Font.Name()
+}
+
+func (f *FontFace) FontFace(dpi float64) font.Face {
+	face, err := opentype.NewFace(f.Face, &opentype.FaceOptions{
+		Size: f.Font.Size.Points(),
+		DPI:  dpi,
+	})
+	if err != nil {
+		panic(err)
+	}
+	return face
+}
+
+// default hinting for OpenType fonts
+const defaultHinting = font.HintingNone
+
+// Extents returns the FontExtents for a font.
+func (f *FontFace) Extents() FontExtents {
+	var (
+		// TODO(sbinet): re-use a Font-level sfnt.Buffer instead?
+		buf  sfnt.Buffer
+		ppem = fixed.Int26_6(f.Face.UnitsPerEm())
+	)
+
+	met, err := f.Face.Metrics(&buf, ppem, defaultHinting)
+	if err != nil {
+		panic(fmt.Errorf("could not extract font extents: %v", err))
+	}
+	scale := f.Font.Size / Points(float64(ppem))
+	return FontExtents{
+		Ascent:  Points(float64(met.Ascent)) * scale,
+		Descent: Points(float64(met.Descent)) * scale,
+		Height:  Points(float64(met.Height)) * scale,
+	}
+}
+
+// Width returns width of a string when drawn using the font.
+func (f *FontFace) Width(s string) Length {
+	var (
+		pixelsPerEm = fixed.Int26_6(f.Face.UnitsPerEm())
+
+		// scale converts sfnt.Unit to float64
+		scale = f.Font.Size / Points(float64(pixelsPerEm))
+
+		width     = 0
+		hasPrev   = false
+		buf       sfnt.Buffer
+		prev, idx sfnt.GlyphIndex
+		hinting   = defaultHinting
+	)
+	for _, rune := range s {
+		var err error
+		idx, err = f.Face.GlyphIndex(&buf, rune)
+		if err != nil {
+			panic(fmt.Errorf("could not get glyph index: %v", err))
+		}
+		if hasPrev {
+			kern, err := f.Face.Kern(&buf, prev, idx, pixelsPerEm, hinting)
+			switch {
+			case err == nil:
+				width += int(kern)
+			case errors.Is(err, sfnt.ErrNotFound):
+				// no-op
+			default:
+				panic(fmt.Errorf("could not get kerning: %v", err))
+			}
+		}
+		adv, err := f.Face.GlyphAdvance(&buf, idx, pixelsPerEm, hinting)
+		if err != nil {
+			panic(fmt.Errorf("could not retrieve glyph's advance: %v", err))
+		}
+		width += int(adv)
+		prev, hasPrev = idx, true
+	}
+	return Points(float64(width)) * scale
+}
+
+// Collection is a collection of fonts, regrouped under a common typeface.
+type Collection []FontFace
+
+// Cache collects font faces.
+type Cache struct {
+	def   Typeface
+	faces map[Font]*opentype.Font
+}
+
+func (c *Cache) GobEncode() ([]byte, error) { return nil, nil }
+func (c *Cache) GobDecode([]byte) error {
+	if c.faces == nil {
+		c.faces = make(map[Font]*opentype.Font)
+	}
+	return nil
+}
```

Comment:
Are these being used?

Reply:
this is to make `plot/gob` happy.
otherwise, `encoding/gob` will try to (de)serialize `opentype.Font` which doesn't have any exported field (and `gob` doesn't like that.)

I'll add some comment about that, there.